### ISSUE TITLE
Imagick 3.4 is throwing exceptions in PHP 7

### ISSUE THREAD

**Issue by ocean90** [mkoppanen/imagick#147]
WordPress uses the error suppress operator for some of the image functions like `exif_read_data()` or `imagecreatefromstring()`. Yesterday we noticed that some of our tests were failing because we're using an image which has no/broken meta data. The tests are failing because of an `ImagickException: exif_read_data(waffles.jpg): Incorrect APP1 Exif Identifier Code` error which was previously just a PHP warning.

I couldn't find a note about this change in the changelog. Can you point me to the related commit please and/or is this related to http://php.net/manual/en/language.errors.php7.php?

See https://core.trac.wordpress.org/ticket/36587 for our ticket.

**Comment by Danack** [comment]
Uhhh, I don't understand what you're asking.

Neither [exif_read_data](http://php.net/manual/en/function.exif-read-data.php) or [imagecreatefromstring](http://php.net/manual/en/function.imagecreatefromstring.php) are anything to do with Imagick.

**Comment by Danack** [comment]
Feel free to provide a reproduce case, but it just looks like your tests have weird output: 

> 89) Tests_Query_Conditionals::test_search_paged
> 835	ImagickException: Cannot modify header information - headers already sent by (output started at /tests/phpunit/includes/bootstrap.php:56)

And are just attaching ImagickException to everything.

**Comment by dd32** [comment]
The debug log from that PHPUnit run had a huge number of issues, but unfortunately it's definitely all caused by the Imagick extension (or potentially a bug within PHP itself). The reason the log was so full was that once a Imagick exception is triggered, every PHP warning thereafter is thrown as a `ImagickException` (where it shouldn't be thrown at all).

Here's a minimal script to reproduce it:
```
<?php
error_reporting( E_ALL ); ini_set( "display_errors", true );

try {
        $f = new Imagick( 'http://any/url/here' );
} catch( Exception $imagick_exception ) {
        var_dump( compact( 'imagick_exception' ) );
}

getimagesize( "something here" );
```

What you'd expect to see here is the `imagick_exception` data followed by a PHP Warning, instead you get the `imagick_exception` data followed by  a PHP Fatal:
```
Fatal error: Uncaught ImagickException: getimagesize(something here): failed to open stream: No such file or directory
```

**Comment by Danack** [comment]
Okay...that would be possible if Imagick was setting the internal PHP error handler, and then not resetting it....which it looks like it's possibly [doing here](https://github.com/mkoppanen/imagick/blob/master/imagick_file.c#L190-L210)

I don't think that code has changed recently - so I'll need to investigate what's changed in PHP to trigger the bad behaviour and/or make sure to restore the error handler.

**Comment by Danack** [comment]
I've committed a fix [for this](https://github.com/mkoppanen/imagick/commit/e95c39427210f709fd8b8bca439e39984f04563e). Probably. The code looks like it has been incorrect for years, so I'm uncertain why it was working before.

Are you able to compile the master version of Imagick to test against that, so that you can confirm that your unit tests are now running correctly?

**Comment by ocean90** [comment]
Compiled the master version via `brew reinstall --HEAD php70-imagick` and tests are running correctly with the fix.

**Comment by Danack** [comment]
Okay - I'll have to do a new release with that fix in, which I'll try to do Saturday.

**Comment by Danack** [comment]
This should be fixed in version 3.4.2 which has a beta release https://pecl.php.net/package/imagick/3.4.2

Which should be installable through `pecl install imagick-3.4.2`

And I forgot to do a RC candidate of it......drat.

**Comment by dd32** [comment]
I've tested `pecl install imagick-3.4.2` on the system I used to test, it's indeed solved the issue.

Thanks @Danack
